### PR TITLE
Added interfaces and tests

### PR DESCRIPTION
**Pull request by XedinUnknown**

### CODE REVIEW

File: src/Exception/NotFoundExceptionInterface.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+namespace Dhii\Config\Exception;
+
+use Dhii\Data\Container\Exception\NotFoundExceptionInterface as BaseNotFoundExceptionInterface;
+
+/**
+ * Represents an exception that occurs in relation to a config.
+ *
+ * @since [*next-version*]
+ */
+interface NotFoundExceptionInterface extends
```
Comment (mecha) [inline src/Exception/NotFoundExceptionInterface.php:12]: Consider allowing retrieval of the key that was not found.

File: src/ConfigInterface.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ */
+interface ConfigInterface extends
+        ChildrenAwareNodeInterface,
+        KeyAwareNodeInterface,
+        RecursiveIteratorInterface,
+        ContainerInterface,
+        Stringable
+{
+    /**
+     * @param string|Stringable $key The config key.
```
Comment (mecha) [inline src/ConfigInterface.php]: Missing an `{@inheritdoc}`. Is this intentional?

**Review by mecha — CHANGES_REQUESTED** [review]

File: src/Exception/NotFoundExceptionInterface.php (hunk 1 above)
Comment (XedinUnknown) [reply src/Exception/NotFoundExceptionInterface.php:12]: `Dhii\Data\Container\Exception\NotFoundExceptionInterface`

File: src/ConfigInterface.php (hunk 2 above)
Comment (XedinUnknown) [reply src/ConfigInterface.php]: Most likely missing

File: src/Exception/NotFoundExceptionInterface.php (hunk 1 above)
Comment (mecha) [reply src/Exception/NotFoundExceptionInterface.php:12]: Oh my bad. I thought it was extended the PSR-11 `NotFoundExceptionInterface`.

**Comment by XedinUnknown** [comment]
@mecha, it appears that right now `getValue()` return type is dependent on the internal storage of the configuration. This should not happen. Can we agree that the value will always be a string? Does this make sense to you? It would solve the problem of consumers having to cast the value twice in many cases.

**Comment by mecha** [comment]
@XedinUnknown Yes. Perhaps also consider `StringableInterface`.

**Comment by XedinUnknown** [comment]
@mecha, you mean [like this](https://github.com/Dhii/config-interface/blob/ef5394a0abc36cb0bab7fc5eb8b20bd56d7e10bb/src/ConfigInterface.php#L23)?

**Comment by mecha** [comment]
@XedinUnknown No, I meant `getValue(): string|StringableInterface`.

**Comment by XedinUnknown** [comment]
In my comment, I explain that `getValue(): string` solves the problem of having to cast prior to any conversion, Allowing it to return `StringableInterface` does not solve that, @mecha.

**Comment by mecha** [comment]
Fair point.

**Comment by XedinUnknown** [comment]
And if what's inside is a hierarchy, convert to JSON?

**Comment by mecha** [comment]
I would favor whatever yields best performance in this case. If the hierarchy consists of plain arrays, JSON encoding and decoding is faster. If the hierarchy contains objects, PHP's serializing and unserializing is faster.

**Comment by XedinUnknown** [comment]
@mecha, because no consumer should care or even be able to know what the structure actually is, going with that logic would make the return format unpredictable.

**Comment by mecha** [comment]
I didn't mean that the format would be determined based on the contents of the hierarchy. Rather, we pick a form of serializing, either PHP's or JSON, based on what is more likely to be the case.

**Comment by XedinUnknown** [comment]
@mecha, I don't think that one of them is more likely. I suggest we pick based on what is more useful. IMHO, that's JSON.

**Comment by mecha** [comment]
@XedinUnknown I think serializing can be more useful. 

1. JSON does not include non-public data, resulting in loss of state
2. Serializing is almost always faster based on benchmarks
3. Objects can choose how to serialize or re-hydrate themselves

**Comment by XedinUnknown** [comment]
@mecha, the result of `getValue()` does not need to know non-public data. In fact, it does not need to know any data that is not key and value data.

**Comment by mecha** [comment]
@XedinUnknown Ok. Then if points 2 and 3 are not a concern, feel free to use JSON.

**Comment by XedinUnknown** [comment]
Speed is certainly a concern. However, it's better to have a useful result slower, rather than a useless result faster. What are you gona do with your serialization result? You're not gona send it to client, or store it in DB, because nowadays JSON is used for that. 

As for point 3, like I wrote, there's no need for objects to re-hydrate themselves, because config consumers don't and shouldn't and cannot care about what kind of objects are inside the config.

**Pull request merged**